code/controllers/subsystem/air.dm

```diff
@@ -415,3 +423,74 @@ SUBSYSTEM_DEF(air)
 		return gas_string
 	var/datum/atmosphere/mix = atmos_gen[gas_string]
 	return mix.gas_string
+
+
+/datum/controller/subsystem/air/ui_state(mob/user)
+	return GLOB.debug_state
+
+/datum/controller/subsystem/air/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "AtmosControlPanel")
+		ui.set_autoupdate(FALSE)
+		ui.open()
+
+/datum/controller/subsystem/air/ui_data(mob/user)
+	var/list/data = list()
+	data["excited_groups"] = list()
+	for(var/datum/excited_group/group in excited_groups)
+		var/turf/T = group.turf_list[1]
+		var/area/target = get_area(T)
+		data["excited_groups"] += list(list(
+			"jump_to" = REF(T), //Just go to the first turf
+			"group" = REF(group),
+			"area" = target.name,
+			"breakdown" = group.breakdown_cooldown,
+			"dismantle" = group.dismantle_cooldown,
+			"size" = group.turf_list.len,
+			"should_show" = group.should_display
+		))
+	data["active_size"] = active_turfs.len
+	data["hotspots_size"] = hotspots.len
+	data["excited_size"] = excited_groups.len
+	data["conducting_size"] = active_super_conductivity.len
+	data["frozen"] = can_fire
+	data["show_all"] = display_all_groups
+	data["fire_count"] = times_fired
+	return data
+
+/datum/controller/subsystem/air/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+	if(..() || !check_rights_for(usr.client, R_DEBUG))
+		return
```

Comment:
```suggestion
	. = ..()
	if(. || !check_rights_for(usr.client, R_DEBUG))
		return
```

Reply:
The parent here is only used to see if it's alright to display/refresh the window, and from what I can tell to prevent shady topic calls. This is appropriate use.

Reply:
Always preserve the parent call return if you are using it.
Example issue:
```
/atom/proc/safe_return()
	if(condition_to_quit)
		return TRUE
	stuff()

/atom/movable/safe_return()
	if(..())
		return
	do_stuff()

/obj/safe_return()
	if(..())
		return
	do_stuff()
```

Parent return preserving is dirt cheap. If on doubt, use it.

Reply:
the parent call for ui_act is used in this manner everywhere I can find. The expected returns of the proc are either TRUE to refresh the ui, or FALSE/null, to not do that. Even if the controller or subsystem returned something related to this, I wouldn't want it. as it stands, the parent call for ui_act returns whether to continue with the check, not whether to refresh the page, so using it in this manner would cause unintended behavior

Reply:
So in this particular case it's using the parent return differently?

Reply:
Yes. The parent return of ui_act is not used to convey the same information as a normal return.